Device handling rework

GammaSQ opened the pull request:
In Q4.0, devices are dealt with very differently than in Q3.2. I initially wanted to write better documentation on just the [`qvm-device`](https://github.com/QubesOS/qubes-doc/pull/746) command, but it soon turned out that even the GUI (with the Device Widget) had substantial changes that necessitated a completely different structure than before.
The new structure describes general device handling in one page, while dealing with PCI, USB and block devices on their own page. Furthermore, "Creating a USB qube" and (security)"Implications of using devices" are their own pages now. (The second I believe is especially important for people trying to get a grasp of the threat-model and therefore reading through the Security Guides, not the device-handling-guide)

Regarding USB and Block device handling, the differences between Q4.0 and Q3.2 were so big, nearly no reference to the other page were necessary. (Except for creating a USB qube, but that's it's own page now.)
However, PCI-handling stayed quite similar, so "Assigning PCI Devices to AppVMs" now has many ugly references to the new "PCI Devices" page.

There is still some work to be done. I've left some comments starting with `<!--TODO:` in the docs. An overview:
 - Icons: I've tried to use Icons when documenting the GUI, however, I was unable to figure out which icon-pack qubes uses. I've referenced [this project's icons](https://github.com/hrdwrrsk/adwaita-xfce-icon-theme/tree/master/Adwaita-Xfce) where possible, but would love some directions where 1.) the Device Manager Widget icon comes from (I use a differently colored one) and 2.) the USB-icon in the Widget comes from (I couldn't find one that looks similar enough.)
 - Security Guide: I'm not a security-guru. Someone please check the `device-considerations.md` file for errors!
 - It seems to be possible to attach the same device to several VMs. I know I did that accidentally with a block device once. Is this a bug? Is it possible with USB devices as well? In` device-handling.md`
 - Does `qvm-block l work` list exclusively devices AVAILABLE IN or also devices ATTACHED TO work? In `device-handling.md`
 - There was a [commit](https://github.com/marmarek/qubes-core-admin-client/commit/32abd8bf25301f7cf4cfe8d47aae003e1774bff0) by @marmarek to detach all devices. Is this part of qubes-updates by now? In `device-handling.md`
 - There is a guide on finding the right USB controller for pci passthrough. It looks like it was written before there was usb-passthrough? Can somebody confirm this may still be required? In `usb-devices.md`
 - Similarly, there is a guide on how to bring back PCI devices to dom0. Is there a usecase? in pci-devices.md
 - The `qubesctl state.sls qvm.usb-keyboard` is mentioned, though not how to undo it. How to remove keyboard-privileges? in `usb-qube-howto.md`
 - In the manual setup of a usb-qube, it wasn't clear to me how a conformation dialogue could be used during LUKS-decryption? Maybe a more obvious distinction is needed here? ("this works with LUKS decryption, and THIS does not!") in `usb-qube-howto.md`
 - similarly, why is `user=root` required in the rpc-policy? I kinda get why a keyboard needs root privileges, but a mouse? in `usb-qube-howto.md`
 - If there is a block device sda with partitions sda1 and sda2, after `qvm-block a work sys-usb:sda`, the partitions are available as `/dev/xvdi1` and `/dev/xvdi2`. However, if I attach only one partition `qvm-block a work sys-usb:sda1`, is it now available as `/dev/xvdi1` or `/dev/xvdi`? in `block-devices.md`
 - For Q3.2 there was a guide on how to handle prematurely detached devices. How is this handled in Q4.0? in `block-devices.md`

marmarek commented on common-tasks/block-devices.md:
```diff
@@ -0,0 +1,168 @@
+where `xvdi2` needs to be replaced with the partition you want to mount.
+This will make your drive content accessible under `~/mnt`.
+
+Beware that when you attach a whole block device, partitions can be identified by their trailing integer (i.e. `/dev/xvdi2` for the second partition, `/dev/xvdi` for the whole device), whereas if you attach a single parition, the partition has *no trailing integer*. <!--TODO: really? Didn't have a drive to test this quickly.-->
```
> TODO: really? Didn't have a drive to test this quickly.

Yes, that's right.

andrewdavidwong commented:
Thanks for doing this, @GammaSQ! As a general note, I've noticed that some of the page organization here diverges from our [documentation guidelines](https://www.qubes-os.org/doc/doc-guidelines/), specifically the part about [version-specific documentation](https://www.qubes-os.org/doc/doc-guidelines/#version-specific-documentation). Just a heads up that some or all of the content that you submit will probably be reorganized in accordance with these guidelines and the current organization of Qubes docs (e.g., multiple pages may be combined into one, and content may be added to an existing page instead of as a new page).

GammaSQ commented on common-tasks/block-devices.md:
```diff
@@ -0,0 +1,168 @@
+#Command Line Tool Guide
+The command-line tool you may use to mount whole USB drives or their partitions is `qvm-block`, a shortcut for `qvm-device block`.
+
+`qvm-block` only sees device-nodes and may not use names you expect! So make sure to have the drive available in the sourceVM before listing available block devices (step 1.) to find out it's its ID.
```
New version:

 > `qvm-block` won't recognise your device by any given name, but rather the device-node the sourceVM assigns. So make sure you have the drive available in the sourceVM, then list the available block devices (step 1.) to find the corresponding device-node.

Is this more clear?

GammaSQ commented on common-tasks/block-devices.md:
```diff
@@ -0,0 +1,168 @@
+        qvm-block detach work sys-usb:sdb
+
+ 6.  You may now remove the device or attach it to another qube.
+<!--TODO: what happens if USB-device removed before detaching? How about attaching a blockdevice to two qubes?-->
```
That part of the documentation is specifically for 3.2, so I ignored it for 4.0. Is there a version for 4.0 or is it equally valid, despite the title?

marmarek replied:
Since [qubes-utils 4.0.6](https://github.com/marmarek/qubes-linux-utils/commit/764b0f3f0778c7675ded7aa67aa954917045f20a) much simpler recovery is possible:
```
virsh detach-disk work xvdi
```
(adjust `work` and `xvdi`)
Unless device is originally from dom0, then the R3.2 procedure needs to be used.

GammaSQ commented on configuration/usb-qube-howto.md:
```diff
@@ -0,0 +1,210 @@
+The above command will take care of all required configuration, including creating USB qube if not present.
+Note that it will expose dom0 to USB devices while entering LUKS passphrase.
+Users are advised to physically disconnect other devices from the system for that time, to minimize the risk.
+<!--TODO: How to undo?-->
```
The question more specifically is about this command: `sudo qubesctl state.sls qvm.usb-keyboard`. Is there no similar undo-command?

marmarek replied:
No, there is no simplified undo.

GammaSQ commented on configuration/usb-qube-howto.md:
```diff
@@ -0,0 +1,210 @@
+
+    sys-usb dom0 allow,user=root
+    
+(Change `sys-usb` to your desired USB qube.) <!--TODO: Why user=root?-->
```
So I can just remove the `,user=root` part?

GammaSQ commented:
@andrewdavidwong You are right, based on guidelines this PR would have to be rejected. However, I kindly ask to judge this PR on its merit.
I didn't see a good way to comply with the guideline and have an easy to grasp, complete documentation in the case of device handling.

GammaSQ commented on common-tasks/pci-devices.md:
```diff
@@ -0,0 +1,148 @@
+
+**Security warning:** PCI passthrough should be safe by default, but none-default options may be required. Please make sure you carefully read and understood the **[security considerations]** before deviating from default behavior!
+
+Unlike other devices ([USB], [block], mic), PCI devices need to be attached on VM-bootup. Similar to how you can't attach a new sound-card after your computer booted (and expect it to work properly), attaching PCI devices to already booted VMs isn't possible in any meaningful way.
```
Actually, I'm not entirely sure what you mean. I'm under the impression that reattaching a PCI device (w/o FLR) may be unsafe, but detaching itself should yield no additional security implications. (There is a paragraph on reattaching in device-considerations.md)

marmarek replied:
There are also cases where detaching PCI device w/o FLR may be unsafe.

marmarek commented on common-tasks/device-handling.md:
```diff
@@ -0,0 +1,118 @@
+To detach a device, click the Qubes Devices Widget icon again. Attached devices are displayed in bold. Hover the one you want to detach. A list of VMs appears, one showing the eject symbol: ![eject icon]
+
+##Attaching a Device to Several VMs
+<!--TODO: This seems like a very bad idea, but it's possible? When would I want to do that? What are the security-implications?-->
```
I couldn't find one, so created here: https://github.com/QubesOS/qubes-issues/issues/4692

andrewdavidwong commented:
> You are right, based on guidelines this PR would have to be rejected. However, I kindly ask to judge this PR on its merit.
I didn't see a good way to comply with the guideline and have an easy to grasp, complete documentation in the case of device handling.

No worries. I wasn't suggested that the PR should be rejected for this reason. I just wanted to let you know that I'll probably reorganize the content as part of the merge in order to bring it in line with the rest of the documentation and the guidelines.

GammaSQ commented:
@marmarek Please have a look especially at the changes in e534772, I'm not 100% sure I understood your requested changes there.

@andrewdavidwong Thank you. Though I'd recommend to keep the structure. Especially having a dedicated page for security-implications seems important. But the idea of having one page for both Q4.0 and Q3.2 doesn't really work if there are so many differences. (Personally, I found the previous usb.md confusing due to it's constant switching between Q4.0 and Q3.2 *and* USB and block devices.)

What's still missing:
 - The Icon for the device-widget
 - More (solid) information on PCI-devices (as discussed in [this comment](https://github.com/QubesOS/qubes-doc/pull/763#discussion_r245488688) and mentioned in my comment in `device-considerations.md`)

andrewdavidwong commented:
> Thank you. Though I'd recommend to keep the structure. Especially having a dedicated page for security-implications seems important. But the idea of having one page for both Q4.0 and Q3.2 doesn't really work if there are so many differences. (Personally, I found the previous usb.md confusing due to it's constant switching between Q4.0 and Q3.2 and USB and block devices.)

Ah, but that's precisely why consistency is important. The constant switching between 4.0 and 3.2 is the exact opposite of what the policy prescribes. (In fact, it's explicitly the "bad" example in the doc guidelines.)

I'll certainly take your recommendation into account, but having many different opinions about the "right" structure for the documentation is exactly what leads to the confusion that prompted you to make many of these changes.

GammaSQ commented on common-tasks/usb-devices.md:
```diff
@@ -0,0 +1,144 @@
+#Attaching And Detaching a USB Device
+##With Qubes Device Manager
+Click the device-manager-icon: ![device manager icon]  
+A list of available devices appears. USB-devices have a USB-icon to their right. <!--TODO: Where is that icon????-->
```
I found that one actually, but I'd need it online somewhere, so I can link it. What repository are we using for icons, so I may link there?

marmarek replied:
There is probably an online version in some XFCE repository or such. But for privacy reasons, better copy it to qubes-attachment repository and use internal link.

marmarek commented on common-tasks/pci-devices.md:
```diff
@@ -0,0 +1,148 @@
+
+Note that one can only attach full PCI or PCI Express devices by default.
+This limit is imposed by the PC and VT-d architectures.
+This means if a PCI device has multiple functions, all instances of it need to be attached to the same qube unless you have disabled the strict requirement with the `no-strict-reset` option during attachment.
```
I'd replace the above with a more generic comment that "for some multifunction devices" it may be needed to assign all of them to the same qube.

marmarek commented:
@GammaSQ I've reviewed remaining comments after your changes. There are still few left. Let me know if I've missed anything.

Also, I believe @andrewdavidwong remark about the structure still stands. Regardless of a valid comment that R3.2 and R4.0 documentation should be on the same page (in two separate sections), I don't think splitting "block" and "usb" devices into separate pages is a good idea. For many people it is very non-obvious that when connecting USB stick, they really should look at _block_ device. Moving that stuff into separate page would make it even harder to realize.
Since "device-handling.md" doesn't have equivalent in R3.2 (and indeed unified framework is something new in R4.0), it can stay as it is. But other R4.0 content should be moved to relevant existing pages.

marmarek commented on common-tasks/device-handling.md:
```diff
@@ -114,6 +114,6 @@ If no specific `sourceVM:deviceID` combination is given, *all devices of that DE
 
 [security considerations]: /doc/device-considerations/
-[device manager icon]: https://raw.githubusercontent.com/hrdwrrsk/adwaita-xfce-icon-theme/master/Adwaita-Xfce/22x22/devices/media-removable.png <!--TODO: find actual icon used in qubes!-->
+[device manager icon]: https://raw.githubusercontent.com/GammaSQ/qubes-attachment/doc-device-rework/wiki/Devices/media-removable.png
```
You should use website's link, not source one. The whole qubes-attachment repository is available at `/attachment/` path, so it will be `/attachment/wiki/Devices/media-removable.png`.

GammaSQ replied:
This existed solely to be independent of the PR to `qubes-attachment`, fixed now.

However, do you want me to replace links to other media with links to `qubes-attachment` as well? e.g. in `pci-devices.md`:

    [appmenu]: https://raw.githubusercontent.com/QubesOS/qubes-artwork/master/icons/32x32/apps/qubes-appmenu-select.png
    [domain manager icon]: https://raw.githubusercontent.com/QubesOS/qubes-artwork/master/icons/32x32/apps/qubes-logo-icon.png

marmarek replied:
Yes, that would be better.

marmarek commented on common-tasks/usb-devices.md:
```diff
@@ -138,7 +138,7 @@ Strip the leading `0000:` and pass the rest to the [`qvm-pci` tool][qvm-pci] to
 [qubes u2f proxy]: /doc/u2f-proxy/
 [4661]: https://github.com/QubesOS/qubes-issues/issues/4661
-[device manager icon]:https://raw.githubusercontent.com/hrdwrrsk/adwaita-xfce-icon-theme/master/Adwaita-Xfce/22x22/devices/media-removable.png <!--TODO: find actual icon used in qubes!-->
+[device manager icon]:https://raw.githubusercontent.com/GammaSQ/qubes-attachment/doc-device-rework/wiki/Devices/media-removable.png
```
And here too.

andrewdavidwong commented:
@GammaSQ, would you be willing to resolve the merge conflict and combine the "block" and "USB" pages as @marmarek suggested in https://github.com/QubesOS/qubes-doc/pull/763#issuecomment-461051410?

GammaSQ commented:
@andrewdavidwong, there are 3 commits which have different levels of merging:
 * af4c726 Just adds a link on top of the USB-documentation explaining people looking for storage-handling should look at the block-device page. (which, btw., is titled "Block or Storage Devices [...]", which I believe makes it clear what the page is about, even if somebody doesn't know what a block device is)
 * 1789ff6 is just placing the block-device explanation before the usb-device explanation, linking happens via `#anchors`.
 * afa8308 is a full merge. I tried to stay topical (first GUI explanation, then command line, then misc), but usb and block devices just behave too differently to make a lot of sense to merge them like this. I strongly discourage this merging, I find it confusing.

Again, I strongly recommend the first version. Having this much information in just one page is too much. It's automatically cluttered. _But_ it's your project, your decision.
Let me know if I can be of any help!

andrewdavidwong commented:
> @andrewdavidwong, there are 3 commits which have different levels of merging:
> 
> * [af4c726](https://github.com/QubesOS/qubes-doc/commit/af4c726b289e08e6c705f5a1c9c4363fc240f36b) Just adds a link on top of the USB-documentation explaining people looking for storage-handling should look at the block-device page. (which, btw., is titled "Block or Storage Devices [...]", which I believe makes it clear what the page is about, even if somebody doesn't know what a block device is)
> * [1789ff6](https://github.com/QubesOS/qubes-doc/commit/1789ff6ddbb16dab4781b16b194fa4fb40af867c) is just placing the block-device explanation before the usb-device explanation, linking happens via `#anchors`.
> * [afa8308](https://github.com/QubesOS/qubes-doc/commit/afa8308c4686bd9fdb39f5b5b11086c0941577cb) is a full merge. I tried to stay topical (first GUI explanation, then command line, then misc), but usb and block devices just behave too differently to make a lot of sense to merge them like this. I strongly discourage this merging, I find it confusing.
> 
> Again, I strongly recommend the first version. Having this much information in just one page is too much. It's automatically cluttered. _But_ it's your project, your decision.
> Let me know if I can be of any help!

@marmarek, what do you think? Since you're the one who suggested keeping "block" and "USB" together, I'd like to know if the third version is what you had in mind.

@GammaSQ, thanks! HTML-Proofer is reporting that this PR results in [28 broken links](https://travis-ci.com/QubesOS/qubes-doc/builds/103117267) (scroll to the bottom). Would you mind fixing those?

marmarek commented:
My main concern is people finding and using "full USB device" attach, when they really need to attach just a block device. Those two have very different attack surface, and since it's equally easy to use them on Qubes 4.0 (which actually might be a mistake...), it's important to know the difference.
Especially when "block device" is not a term you associate with USB stick...
Linking to block devices doc from USB page may be good enough, if that's also clearly marked on documentation index (again, "block device" may be obscure term for people looking for "USB stick") and generally it's easy to land on block devices page when you look for how to use USB stick.
@andrewdavidwong what do you think?

Some long term solution would be to propose in GUI only the "most secure" version by default, or clearly mark which one is such (hiding others in some submenu or such). But right now, no part of devices API know any correlation between instances of different types (like block and usb) which are in fact the same device, so we can't do that right now.
This is something I hope to fix in R4.1, but we'll see whether it will happen in practice.

GammaSQ commented:
> My main concern is people finding and using "full USB device" attach, when they really need to attach just a block device. Those two have very different attack surface, and since it's equally easy to use them on Qubes 4.0 (which actually might be a mistake...), it's important to know the difference.

I agree. That was my motivation in linking the security-considerations-page at the beginning of every page. However, since it actually didn't clarify the block v. usb issue, I added a brief explanation at the beginning of `usb-devices.md`.

> @GammaSQ, thanks! HTML-Proofer is reporting that this PR results in [28 broken links](https://travis-ci.com/QubesOS/qubes-doc/builds/103117267) (scroll to the bottom). Would you mind fixing those?

TBH I don't know how? I though they were detected erroneously, because the new files (`/doc/usb-devices`, `/doc/pci-devices`, aso ...) didn't show up in the diffing-paragraph. Do they need to be added to some kind of makefile?  
I tripple-checked the anchors, corrected some of them in the latest commit, they should be fine.

marmarek commented:
Other pages have links to specific anchors on the old pages, which are now changed. For example see https://qubes-os.org/doc/yubi-key links to /doc/usb/#security-warning-about-usb-input-devices. Travis provide you the full list.

GammaSQ commented:
Indeed, there are some of those cases, and they are now fixed. But most errors claim an anchor ("hash" in the logs) I created does not exist. So I'm expecting I'm missing some kind of "turn-headlines-into-anchors" flag?

marmarek commented:
This is because they are not rendered as headlines. I think you need to add a space after `##`.

andrewdavidwong commented:
> This is because they are not rendered as headlines. I think you need to add a space after `##`.

Yes. Please make sure to test your changes by serving the site locally on your machine and seeing how it renders, @GammaSQ. Instructions to do so are here: https://github.com/QubesOS/qubesos.github.io#instructions

andrewdavidwong commented:
> My main concern is people finding and using "full USB device" attach, when they really need to attach just a block device. Those two have very different attack surface, and since it's equally easy to use them on Qubes 4.0 (which actually might be a mistake...), it's important to know the difference.
> Especially when "block device" is not a term you associate with USB stick...
> Linking to block devices doc from USB page may be good enough, if that's also clearly marked on documentation index (again, "block device" may be obscure term for people looking for "USB stick") and generally it's easy to land on block devices page when you look for how to use USB stick.
> @andrewdavidwong what do you think?

I think that this is something where we shouldn't rely on people reading the documentation, since most users won't read it, and many who do read it will not read carefully enough or still not understand. We should choose safe defaults (which might mean automatically connecting as a block device rather than a full USB device whenever possible) and/or making this distinction discoverable through the OS.

GammaSQ commented:
Finally came around to build qubesos.github.io, links work now.

Needless to say, I'd take care of required fixes if af4c726 is merged.

andrewdavidwong commented:
@GammaSQ: Thanks!

> Needless to say, I'd take care of required fixes if af4c726 is merged.

I'm not sure which required fixes you're referring to here.

@marmarek, not sure if you want to re-review, since there have been changes, but you can now if you like.

GammaSQ commented:
> > Needless to say, I'd take care of required fixes if [`af4c726`](https://github.com/QubesOS/qubes-doc/commit/af4c726b289e08e6c705f5a1c9c4363fc240f36b) is merged.
> 
> I'm not sure which required fixes you're referring to here.

af4c726 is the first of the three commits I offered as choices. (The one which simply links `block-devices.md` atop `usb-devices.md`, which I believe is the best option.)  
Most recent changes include fixing merge-conflicts, typos and links. I would repeat those fixes based on af4c726 if desired.

marmarek commented:
> Most recent changes include fixing merge-conflicts, typos and links. I would repeat those fixes based on [af4c726](https://github.com/QubesOS/qubes-doc/commit/af4c726b289e08e6c705f5a1c9c4363fc240f36b) if desired.

Yes, please do. Other than that, it looks good now.

andrewdavidwong commented:
I'm a little confused, because it looks like https://github.com/QubesOS/qubes-doc/pull/763/commits/6d6f32f7889a032f4b0e0c0b29ae8d69ce00bb52 has new substantive changes, not just fixing merge conflicts, typos, and links.

-----

@marmarek, which one of the three "options" do you think should be merged?

andrewdavidwong commented:
> @marmarek, which one of the three "options" do you think should be merged?

Actually, I may also be confused about this part. Now that the fixes have been committed on top of af4c726, is the idea that the whole PR gets merged? IMHO, having three different "branches" that aren't really branches in the same PR makes things unnecessarily confusing.

marmarek commented:
> @marmarek, which one of the three "options" do you think should be merged?

The first one.

andrewdavidwong commented:
> > Most recent changes include fixing merge-conflicts, typos and links. I would repeat those fixes based on [af4c726](https://github.com/QubesOS/qubes-doc/commit/af4c726b289e08e6c705f5a1c9c4363fc240f36b) if desired.
> 
> Yes, please do. Other than that, it looks good now.



> > @marmarek, which one of the three "options" do you think should be merged?
> 
> The first one.

Ok. I'll wait for @GammaSQ to implement the fixes on top of https://github.com/QubesOS/qubes-doc/commit/af4c726b289e08e6c705f5a1c9c4363fc240f36b then.

GammaSQ commented:
> Ok. I'll wait for @GammaSQ to implement the fixes on top of [af4c726](https://github.com/QubesOS/qubes-doc/commit/af4c726b289e08e6c705f5a1c9c4363fc240f36b) then.

That's the current state of this PR. You can check by `git diff af4c726 common-tasks/block-devices.md` (As that file was removed in the second and third suggestion.)

Travis only complains about words it doesn't know. Should be ready to merge.

andrewdavidwong commented:
> > Ok. I'll wait for @GammaSQ to implement the fixes on top of [af4c726](https://github.com/QubesOS/qubes-doc/commit/af4c726b289e08e6c705f5a1c9c4363fc240f36b) then.
> 
> That's the current state of this PR. You can check by `git diff af4c726 common-tasks/block-devices.md` (As that file was removed in the second and third suggestion.)
> 
> Travis only complains about words it doesn't know. Should be ready to merge.

Oh, ok. Thanks!
